src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp

```diff
@@ -682,106 +662,58 @@ std::optional<DataModel::AttributeInfo> CodegenDataModelProvider::GetAttributeIn
 {
-    std::optional<CommandId> commandId =
-        EnumeratorCommandFinder(&CommandHandlerInterface::EnumerateGeneratedCommands)
-            .FindCommandId(EnumeratorCommandFinder::Operation::kFindFirst,
-                           ConcreteCommandPath(path.mEndpointId, path.mClusterId, kInvalidCommandId));
-    if (commandId.has_value())
-    {
-        return *commandId == kInvalidCommandId ? kInvalidCommandPath
-                                               : ConcreteCommandPath(path.mEndpointId, path.mClusterId, *commandId);
-    }
+    EnumeratorCommandFinder handlerFinder(&CommandHandlerInterface::EnumerateGeneratedCommands);
+    CommandId commandId =
+        FindCommand(ConcreteCommandPath(path.mEndpointId, path.mClusterId, kInvalidCommandId), handlerFinder,
+                    detail::EnumeratorCommandFinder::Operation::kFindFirst, mGeneratedCommandsIterator, GeneratedCommands);
 
-    const EmberAfCluster * cluster = FindServerCluster(path);
-    VerifyOrReturnValue(cluster != nullptr, kInvalidCommandPath);
-
-    commandId = mGeneratedCommandsIterator.First(cluster->generatedCommandList);
-    VerifyOrReturnValue(commandId.has_value(), kInvalidCommandPath);
-    return ConcreteCommandPath(path.mEndpointId, path.mClusterId, *commandId);
+    VerifyOrReturnValue(commandId != kInvalidCommandId, kInvalidCommandPath);
+    return ConcreteCommandPath(path.mEndpointId, path.mClusterId, commandId);
 }
 
 ConcreteCommandPath CodegenDataModelProvider::NextGeneratedCommand(const ConcreteCommandPath & before)
 {
-    // TODO: `Next` redirecting to a callback is slow O(n^2).
-    //       see https://github.com/project-chip/connectedhomeip/issues/35790
-    auto nextId = EnumeratorCommandFinder(&CommandHandlerInterface::EnumerateGeneratedCommands)
-                      .FindCommandId(EnumeratorCommandFinder::Operation::kFindNext, before);
-
-    if (nextId.has_value())
-    {
-        return (*nextId == kInvalidCommandId) ? kInvalidCommandPath
-                                              : ConcreteCommandPath(before.mEndpointId, before.mClusterId, *nextId);
-    }
-
-    const EmberAfCluster * cluster = FindServerCluster(before);
-
-    VerifyOrReturnValue(cluster != nullptr, kInvalidCommandPath);
+    EnumeratorCommandFinder handlerFinder(&CommandHandlerInterface::EnumerateGeneratedCommands);
 
-    std::optional<CommandId> commandId = mGeneratedCommandsIterator.Next(cluster->generatedCommandList, before.mCommandId);
-    VerifyOrReturnValue(commandId.has_value(), kInvalidCommandPath);
+    CommandId commandId = FindCommand(before, handlerFinder, detail::EnumeratorCommandFinder::Operation::kFindNext,
+                                      mGeneratedCommandsIterator, GeneratedCommands);
 
-    return ConcreteCommandPath(before.mEndpointId, before.mClusterId, *commandId);
+    VerifyOrReturnValue(commandId != kInvalidCommandId, kInvalidCommandPath);
+    return ConcreteCommandPath(before.mEndpointId, before.mClusterId, commandId);
```

Comment:
Followup: Doing the return-by-value of the whole path, not just the command id which is the one thing caller does not have, seems a bit space-wasteful too....

Reply:
I can take a look. When I had both command and path versions it was obviously smaller, so I picked the commandid return because this is what "generated commands" needs. That was obviously larger than a single method, hence this updated code.

Reply:
I guess my point is: why do we need to return the whole path at all?

Reply:
It made the API easier to iterate through. All other outputs have a path embedded even though they realistically do not need it and only the ID is needed.

None of the APIs recurse, so they would not really need paths, however ergonomically having full "path + data" seems clearer than "id without prefix + data".